feat: dorq wf list

BenjaminMummery opened the pull request:
# The problem

Users need to be able to see a filtered list of workflow runs via the cli.

# This PR's solution

- user facing:
  - Implements `dorq wf list` with the following arguments
    - `-c` / `--config` specifies a config for which workflows will be listed. Can be specified multiple times. If omitted, the user is prompted to select from a checklist of configs.
    - `-a` / `--all` show all workflow runs that match the specified filters. incompatible with `-i`.
    - `-i` / `--interactive` prompt the user for any filters that are not specified. Incompatible with `-a`.
    - filters:
      - filters are applied per config, so if the user specifies 2 configs and a limit of 10, up to 20 workflow runs will be displayed.
      - `-l` / `--limit` set the maximum number of runs to display per config.
      - `-t` / `--max-age` set the maximum age of runs to display.
      - `-s` / `--state` only show workflow runs of the specified state. Can be specified multiple times. If one or more specified states are not valid workflow run states, prompts the user to choose from available states.
- non-user facing:
  - extends prompter to allow prompting for checkboxes, and integer inputs.
  - refactors workflowrunrepo to make use of the API `list_workflow_runs` method.
  - extends workflowrunrepo to provide `list_wf_runs` method. The preexisting `list_wf_run_ids` method now uses this.

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`

[ORQSDK-671]


[ORQSDK-671]: https://zapatacomputing.atlassian.net/browse/ORQSDK-671?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

BenjaminMummery committed:
feat: add checkbox option to prompter

[ORQSDK-671]

BenjaminMummery committed:
feat: add list command to cli

[ORQSDK-671]

BenjaminMummery committed:
feat: add arg resolvers for list args

[ORQSDK-671]

BenjaminMummery committed:
feat: add arg resolvers for list args

[ORQSDK-671]

BenjaminMummery committed:
feat: add list glue code

[ORQSDK-671]

BenjaminMummery committed:
feat: add list glue code

[ORQSDK-671]

BenjaminMummery committed:
refactor: make dorq use API for listing configs rather than its own implementation

[ORQSDK-671]

BenjaminMummery committed:
feat: add show_wf_runs_list to presenter

[ORQSDK-671]

BenjaminMummery committed:
feat: add help strings for list arguments

[ORQSDK-671]

BenjaminMummery committed:
style: typing

[ORQSDK-671]

BenjaminMummery committed:
fix: convert API workflowRun to schema worklfowrun

[ORQSDK-671]

BenjaminMummery committed:
fix: conversion between schema and api workflowRun

[ORQSDK-671]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-671-orq-wf-list

BenjaminMummery committed:
chore: fix typo

[ORQSDK-671]

BenjaminMummery committed:
docs: update changelog

[ORQSDK-671]

jamesclark-Zapata commented on src/orquestra/sdk/_base/cli/_dorq/_entry.py:
```diff
@@ -154,6 +154,51 @@ def stop(wf_run_id: t.Optional[str], config: t.Optional[str]):
+    "--all",
+    is_flag=True,
+    flag_value=True,
+    help="Show all workflow runs that match the specified filters.",
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** ‏I'm not sure what this help string means

BenjaminMummery replied:
Updated in [e7fc2e5](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/14/commits/e7fc2e5008a07912ef7a116f96de3c53312ee6c4): "Assume that any unset filters are not required."

jamesclark-Zapata replied:
I'm not sure I get why this option is needed, can you help me understand?
If I do `dorq wf list --state RUNNING` I would expect to list the running workflows, but from this help message, it looks like I would need to write `dorq wf list --state RUNNING --all` which looks confusing.

BenjaminMummery replied:
There was a logic train that led here, but on second viewing everything we wanted to do is covered by `-h`. The all option is removed in [865560e](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/14/commits/865560e212714d574dd04dee1d22a407f25c2abd)

jamesclark-Zapata commented on tests/cli/dorq/test_arg_resolvers.py:
```diff
@@ -296,3 +395,217 @@ def test_service_resolver(args, expected_services):
+        @staticmethod
+        def test_passing_limit_directly():
+            # Given
+            limit = 1701
```
Boldly go

jamesclark-Zapata commented on tests/cli/dorq/test_arg_resolvers.py:
```diff
@@ -296,3 +395,217 @@ def test_service_resolver(args, expected_services):
+            # Given
+            prompter = Mock()
+            resolver = _arg_resolvers.WFRunFilterResolver(prompter=prompter)
+            selected_limit = 1649
```
you might need to explain this one :)

jamesclark-Zapata commented on tests/cli/dorq/test_repos.py:
```diff
@@ -110,14 +111,10 @@ def test_passing_errors(monkeypatch, exc):
 
-                runtime = Mock()
-                runtime.get_all_workflow_runs_status.side_effect = exc
-                monkeypatch.setattr(
-                    _factory, "build_runtime_from_config", Mock(return_value=runtime)
-                )
+                monkeypatch.setattr(sdk, "list_workflow_runs", raise_exc)
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion (non-blocking):** I dont have any strong feelings, but I think this could be:
 ‏
```suggestion
                monkeypatch.setattr(sdk, "list_workflow_runs", Mock(side_effect=exc))
```

jamesclark-Zapata reviewed (changes requested):
Looks good, I really like the interactive approach too.

I have a few comments.

jamesclark-Zapata committed:
chore: update arg resolvers and prompters (#20)

# The problem

In some cases, we need to list workflow runs in the CLI and then use the
workflow run.

Previously, we listed the IDs (by getting the runs and then printing the
IDs) and then calling the runtime to get the actual workflow run.

If a user cancelled a request in an arg resolver, we would get an ugly
error message.

# This PR's solution

 - Adds a new resolver to either resolve an ID or a run
- Adds an exception for when a user cancels the prompting in the arg
resolvers.

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new
features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a
user-readable description of the changes (if the change isn't visible to
the user in any way, check the box).
- [x] The PR's title is prefixed with
`<feat/fix/chore/internal/docs>[!]:`

BenjaminMummery committed:
docs: remove erronious numbering.

[ORQSDK-671]

BenjaminMummery committed:
fix: allow user to specify None when asked for integer

[ORQSDK-671]

BenjaminMummery committed:
feat: add prompter.ask_for_str

[ORQSDK-671]

BenjaminMummery commented on src/orquestra/sdk/_base/cli/_dorq/_arg_resolvers.py:
```diff
@@ -39,6 +39,27 @@ def resolve(self, config: t.Optional[ConfigName]) -> ConfigName:
+        if configs is not None and len(configs) > 0:
+            return configs
+
+        # 1.2 Prompt the user.
```
Good point, that's a holdover from me copying the structure of `resolve`. Removed in [9f6c4c9](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/14/commits/9f6c4c901021d39e0bb9a2880fbbd5a0ca8e67d8)

BenjaminMummery commented on src/orquestra/sdk/_base/cli/_dorq/_arg_resolvers.py:
```diff
@@ -131,3 +152,100 @@ def resolve(
+                    "If 'None', all results will be displayed."
+                ),
+                default=None,
+            )
```
Fixed in [12d9f73](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/14/commits/12d9f73905d4c06fb32ca514584bb831e0ceb18d), "None" or an empty string should now parse properly.

BenjaminMummery commented on src/orquestra/sdk/_base/cli/_dorq/_arg_resolvers.py:
```diff
@@ -131,3 +152,100 @@ def resolve(
+            return max_age
+
+        if interactive:
+            return self._prompter.ask_for_str(
```
Well that would present a problem. [1631cf2](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/14/commits/1631cf28254dd613c1a874c5274d37616f588d0f)

BenjaminMummery committed:
docs: add clearer help string for wf list -a

[ORQSDK-671]

BenjaminMummery committed:
feat: remove  option

[ORQSDK-671]

BenjaminMummery committed:
test: update tests for removed 'all' option

[ORQSDK-671]

BenjaminMummery committed:
refactor: cleaner mocking for list_workflow_runs

[ORQSDK-671]

BenjaminMummery committed:
style: fix return type for ask_for_str

[ORQSDK-671]

BenjaminMummery committed:
fix: remove unused option from makefile

[ORQSDK-671]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-671-orq-wf-list

BenjaminMummery committed:
style: reformat imports

[ORQSDK-671]

BenjaminMummery committed:
fix: correct errors caused by merge

[ORQSDK-671]

BenjaminMummery committed:
style: fix import ordering

[ORQSDK-671]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-671-orq-wf-list

BenjaminMummery committed:
fix: correct merge conflicts

[ORQSDK-671]
